File: python/src/pocketsensor/frames.py
"""ARKit の姿勢を REP-103 へ付け替える基準実装。"""

from __future__ import annotations

import math
from typing import Final

import numpy as np
from numpy.typing import ArrayLike, NDArray

ARKIT_TO_REP103: Final[NDArray[np.float64]] = np.array(
    [
        [0.0, 0.0, -1.0],
        [-1.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
    ],
    dtype=np.float64,
)

LINK_TO_COLOR_OPTICAL_RPY: Final[tuple[float, float, float]] = (-math.pi / 2, 0.0, -math.pi / 2)
LINK_TO_IMU_RPY: Final[tuple[float, float, float]] = (0.0, -math.pi / 2, 0.0)


def _canonicalize_quat(x: float, y: float, z: float, w: float) -> NDArray[np.float64]:
    """単位四元数の符号を揃える。w >= 0。w が 0 なら (x, y, z) の最初の非零が正。"""
    n = math.sqrt(x * x + y * y + z * z + w * w)
    x, y, z, w = x / n, y / n, z / n, w / n
    if w < 0.0:
        x, y, z, w = -x, -y, -z, -w
    elif w == 0.0:
        if x < 0.0 or (x == 0.0 and y < 0.0) or (x == 0.0 and y == 0.0 and z < 0.0):
            x, y, z, w = -x, -y, -z, -w
    return np.array([x, y, z, w], dtype=np.float64)


def matrix_to_quat(m: ArrayLike) -> NDArray[np.float64]:
    """3x3 回転行列を (x, y, z, w) にする。Shepperd 法。"""
    m = np.asarray(m, dtype=np.float64)
    trace = float(m[0, 0] + m[1, 1] + m[2, 2])
    if trace > 0:
        s = math.sqrt(trace + 1.0) * 2.0
        w = 0.25 * s
        x = (m[2, 1] - m[1, 2]) / s
        y = (m[0, 2] - m[2, 0]) / s
        z = (m[1, 0] - m[0, 1]) / s
    elif m[0, 0] > m[1, 1] and m[0, 0] > m[2, 2]:
        s = math.sqrt(1.0 + m[0, 0] - m[1, 1] - m[2, 2]) * 2.0
        w = (m[2, 1] - m[1, 2]) / s
        x = 0.25 * s
        y = (m[0, 1] + m[1, 0]) / s
        z = (m[0, 2] + m[2, 0]) / s
    elif m[1, 1] > m[2, 2]:
        s = math.sqrt(1.0 + m[1, 1] - m[0, 0] - m[2, 2]) * 2.0
        w = (m[0, 2] - m[2, 0]) / s
        x = (m[0, 1] + m[1, 0]) / s
        y = 0.25 * s
        z = (m[1, 2] + m[2, 1]) / s
    else:
        s = math.sqrt(1.0 + m[2, 2] - m[0, 0] - m[1, 1]) * 2.0
        w = (m[1, 0] - m[0, 1]) / s
        x = (m[0, 2] + m[2, 0]) / s
        y = (m[1, 2] + m[2, 1]) / s
        z = 0.25 * s
    return _canonicalize_quat(float(x), float(y), float(z), float(w))


def quat_to_matrix(q: ArrayLike) -> NDArray[np.float64]:
    x, y, z, w = (float(v) for v in np.asarray(q, dtype=np.float64).reshape(4))
    n = math.sqrt(x * x + y * y + z * z + w * w)
    x, y, z, w = x / n, y / n, z / n, w / n
    return np.array(
        [
            [1 - 2 * (y * y + z * z), 2 * (x * y - z * w), 2 * (x * z + y * w)],
            [2 * (x * y + z * w), 1 - 2 * (x * x + z * z), 2 * (y * z - x * w)],
            [2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x * x + y * y)],
        ],
        dtype=np.float64,
    )


def quaternion_to_yaw(orientation_xyzw: ArrayLike) -> float:
    """ROS の yaw（rad）。R = Rz(yaw) Ry(pitch) Rx(roll) なので、行列の xy から取る。"""
    r = quat_to_matrix(orientation_xyzw)
    return math.atan2(float(r[1, 0]), float(r[0, 0]))


def rotate_vector(orientation_xyzw: ArrayLike, vector: ArrayLike) -> NDArray[np.float64]:
    """四元数が表す回転をベクトルへ掛ける。vector は (3,) か (N, 3)。"""
    r = quat_to_matrix(orientation_xyzw)
    v = np.asarray(vector, dtype=np.float64)
    if v.ndim == 1:
        return r @ v.reshape(3)
    return (r @ v.reshape(-1, 3).T).T


def rpy_to_quaternion(roll: float, pitch: float, yaw: float) -> NDArray[np.float64]:
    """ROS の固定軸 RPY。R = Rz(yaw) Ry(pitch) Rx(roll)。返り値は (x, y, z, w)。"""
    cr = math.cos(roll)
    sr = math.sin(roll)
    cp = math.cos(pitch)
    sp = math.sin(pitch)
    cy = math.cos(yaw)
    sy = math.sin(yaw)
    rx = np.array([[1.0, 0.0, 0.0], [0.0, cr, -sr], [0.0, sr, cr]], dtype=np.float64)
    ry = np.array([[cp, 0.0, sp], [0.0, 1.0, 0.0], [-sp, 0.0, cp]], dtype=np.float64)
    rz = np.array([[cy, -sy, 0.0], [sy, cy, 0.0], [0.0, 0.0, 1.0]], dtype=np.float64)
    return matrix_to_quat(rz @ ry @ rx)


def relative_pose(
    parent_position: ArrayLike,
    parent_orientation_xyzw: ArrayLike,
    child_position: ArrayLike,
    child_orientation_xyzw: ArrayLike,
) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
    """同じ frame で表した親と子の姿勢から、親の frame で見た子の位置と四元数を求める。

    端末の姿勢（odom から link）と anchor（odom から anchor）を渡すと、端末から見た anchor になる。
    2 つは同じ時刻のものを渡す。時刻がずれると、そのあいだの端末の動きが誤差として入る。
    """
    r_parent = quat_to_matrix(parent_orientation_xyzw)
    r_child = quat_to_matrix(child_orientation_xyzw)
    delta = np.asarray(child_position, dtype=np.float64).reshape(3) - np.asarray(
        parent_position, dtype=np.float64
    ).reshape(3)
    return r_parent.T @ delta, matrix_to_quat(r_parent.T @ r_child)


def arkit_pose_to_rep103(transform: ArrayLike) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
    """ARKit の camera-to-world（行優先 4x4）を REP-103 の位置と四元数にする。"""
    t = np.asarray(transform, dtype=np.float64).reshape(4, 4)
    r_cam = t[:3, :3]
    p_arkit = t[:3, 3]
    r = ARKIT_TO_REP103
    position = r @ p_arkit
    r_out = r @ r_cam @ r.T
    return position.astype(np.float64, copy=False), matrix_to_quat(r_out)

File: ios/PocketSensorKit/Sources/PocketSensorCore/Conversion/Frames.swift
import simd

/// ARKit の姿勢を REP-103 へ付け替える。行列の計算は Python の基準実装と同じ手順。
public enum Frames {
    /// 行は [0, 0, -1], [-1, 0, 0], [0, 1, 0]。simd は列優先なので rows: で与える。
    public static let arkitToREP103 = simd_double3x3(rows: [
        SIMD3(0, 0, -1),
        SIMD3(-1, 0, 0),
        SIMD3(0, 1, 0),
    ])

    /// 光学 frame への固定回転。R = Rz(-π/2) Ry(0) Rx(-π/2)。
    public static let linkToColorOptical = quaternion(roll: -.pi / 2, pitch: 0, yaw: -.pi / 2)

    /// IMU frame への固定回転。R = Rz(0) Ry(-π/2) Rx(0)。並進は未較正のため 0。
    public static let linkToImu = quaternion(roll: 0, pitch: -.pi / 2, yaw: 0)

    public static func arkitPoseToREP103(
        _ cameraTransform: simd_double4x4
    ) -> (position: SIMD3<Double>, orientation: simd_quatd) {
        let r = arkitToREP103
        let pArkit = SIMD3(cameraTransform.columns.3.x, cameraTransform.columns.3.y, cameraTransform.columns.3.z)
        let rCam = simd_double3x3(
            SIMD3(cameraTransform.columns.0.x, cameraTransform.columns.0.y, cameraTransform.columns.0.z),
            SIMD3(cameraTransform.columns.1.x, cameraTransform.columns.1.y, cameraTransform.columns.1.z),
            SIMD3(cameraTransform.columns.2.x, cameraTransform.columns.2.y, cameraTransform.columns.2.z)
        )
        let position = r * pArkit
        let rOut = r * rCam * r.transpose
        return (position, canonical(quaternion(from: rOut)))
    }

    /// 単位四元数の符号を揃える。w >= 0。w が 0 なら (x, y, z) の最初の非零が正。
    public static func canonical(_ q: simd_quatd) -> simd_quatd {
        let n = simd_length(q.vector)
        let v = n == 0 ? q.vector : q.vector / n
        if v.w < 0 {
            return simd_quatd(vector: -v)
        }
        if v.w == 0 {
            if v.x < 0 || (v.x == 0 && v.y < 0) || (v.x == 0 && v.y == 0 && v.z < 0) {
                return simd_quatd(vector: -v)
            }
        }
        return simd_quatd(vector: v)
    }

    /// ROS の固定軸 RPY を度で返す。R = Rz(yaw) Ry(pitch) Rx(roll)。
    /// yaw は (-180, 180]。pitch は ±90 で打ち切る。ジンバルロックでも NaN は出さない。
    public static func rpyDegrees(from q: simd_quatd) -> (roll: Double, pitch: Double, yaw: Double) {
        let n = canonical(q)
        let x = n.vector.x
        let y = n.vector.y
        let z = n.vector.z
        let w = n.vector.w
        let sinp = 2.0 * (w * y - z * x)
        let pitchRad: Double
        if sinp >= 1.0 {
            pitchRad = .pi / 2
        } else if sinp <= -1.0 {
            pitchRad = -.pi / 2
        } else {
            pitchRad = asin(sinp)
        }
        let rollRad = atan2(2.0 * (w * x + y * z), 1.0 - 2.0 * (x * x + y * y))
        let yawRad = atan2(2.0 * (w * z + x * y), 1.0 - 2.0 * (y * y + z * z))
        let toDeg = 180.0 / Double.pi
        return (
            roll: Units.wrapToPi(rollRad) * toDeg,
            pitch: pitchRad * toDeg,
            yaw: Units.wrapToPi(yawRad) * toDeg
        )
    }

    /// ROS の固定軸 RPY。R = Rz(yaw) Ry(pitch) Rx(roll)。
    public static func quaternion(roll: Double, pitch: Double, yaw: Double) -> simd_quatd {
        let cr = cos(roll)
        let sr = sin(roll)
        let cp = cos(pitch)
        let sp = sin(pitch)
        let cy = cos(yaw)
        let sy = sin(yaw)
        let rx = simd_double3x3(rows: [
            SIMD3(1, 0, 0),
            SIMD3(0, cr, -sr),
            SIMD3(0, sr, cr),
        ])
        let ry = simd_double3x3(rows: [
            SIMD3(cp, 0, sp),
            SIMD3(0, 1, 0),
            SIMD3(-sp, 0, cp),
        ])
        let rz = simd_double3x3(rows: [
            SIMD3(cy, -sy, 0),
            SIMD3(sy, cy, 0),
            SIMD3(0, 0, 1),
        ])
        return quaternion(from: rz * ry * rx)
    }

    /// 3x3 回転行列を四元数にする。Shepperd 法。Python の matrix_to_quat と同じ分岐。
    public static func quaternion(from m: simd_double3x3) -> simd_quatd {
        func at(_ row: Int, _ col: Int) -> Double {
            m[col][row]
        }

        let trace = at(0, 0) + at(1, 1) + at(2, 2)
        let x: Double
        let y: Double
        let z: Double
        let w: Double
        if trace > 0 {
            let s = sqrt(trace + 1.0) * 2.0
            w = 0.25 * s
            x = (at(2, 1) - at(1, 2)) / s
            y = (at(0, 2) - at(2, 0)) / s
            z = (at(1, 0) - at(0, 1)) / s
        } else if at(0, 0) > at(1, 1), at(0, 0) > at(2, 2) {
            let s = sqrt(1.0 + at(0, 0) - at(1, 1) - at(2, 2)) * 2.0
            w = (at(2, 1) - at(1, 2)) / s
            x = 0.25 * s
            y = (at(0, 1) + at(1, 0)) / s
            z = (at(0, 2) + at(2, 0)) / s
        } else if at(1, 1) > at(2, 2) {
            let s = sqrt(1.0 + at(1, 1) - at(0, 0) - at(2, 2)) * 2.0
            w = (at(0, 2) - at(2, 0)) / s
            x = (at(0, 1) + at(1, 0)) / s
            y = 0.25 * s
            z = (at(1, 2) + at(2, 1)) / s
        } else {
            let s = sqrt(1.0 + at(2, 2) - at(0, 0) - at(1, 1)) * 2.0
            w = (at(1, 0) - at(0, 1)) / s
            x = (at(0, 2) + at(2, 0)) / s
            y = (at(1, 2) + at(2, 1)) / s
            z = 0.25 * s
        }
        return canonical(simd_quatd(ix: x, iy: y, iz: z, r: w))
    }
}
